### PR TITLE
fix threafing bug with send buffer and threaded workers

### LINKED ISSUE #334: [4.0.0] UnicodeDecodeError: 'utf-8' codec can't decode byte 0x96 in position 91: invalid start byte
Django 1.11.16
APM agent 4.0.0
APM server 6.5

Exception is raised at https://github.com/elastic/apm-agent-python/blob/f600230a1643a9442605fce959946790c85c4c61/elasticapm/transport/http.py#L62

For my case `body` is the bytes string below:

    b'data read error: invalid character \'\\x00\' looking for beginning of value [\x00\x00P\x02\x00\x00\.....

### PR DESCRIPTION
**Pull request by beniwohli**
When using threaded workers (e.g. Django runserver in Django 2.0+),
writes to the send buffer need to be guarded by a threading lock.
Otherwise we end up with garbled content in the buffer, which then later
on leads to parsing errors in the APM Server.

fixes #334

### CODE REVIEW

File: elasticapm/transport/base.py [hunk 1]
```diff
@@ -53,12 +53,13 @@ def __init__(
...
+        self._queue_lock = threading.Lock()
         self._last_flush = timeit.default_timer()
         self._flush_timer = None
 
     def queue(self, event_type, data, flush=False):
-        self._queue(self.queued_data, {event_type: data})
+        with self._queue_lock:
+            self.queued_data.write((self._json_serializer({event_type: data}) + "\n").encode("utf-8"))
         since_last_flush = timeit.default_timer() - self._last_flush
         queue_size = self.queued_data_size
```
Comment (roncohen) [inline elasticapm/transport/base.py]: `self.queued_data_size` calls `self.queued_data` so calls to `self.queued_data_size` must also be protected by the lock, or you lock inside `self.queued_data`

Comment (beniwohli) [reply elasticapm/transport/base.py]: Great catch. I'll remove the call to `self.queued_data` in `queued_data_size` and instead check for existence of `self._queued_data` directly and return `0` if it doesn't exist. Then we don't need a lock as it is read only. Makes sense?

Comment (simitt) [reply elasticapm/transport/base.py]: You could do something like:
```
with self._queue_lock:
            self.queued_data.write((self._json_serializer({event_type: data}) + "\n").encode("utf-8"))
            queue_size = self.queued_size

```

Comment (beniwohli) [reply elasticapm/transport/base.py]: @roncohen @simitt like this af2ffd490bc

Comment (roncohen) [reply elasticapm/transport/base.py]: yeah that would probably work. It seems like `queue_size` here could end up being the size of a queue that was just flushed and we would then flush again, but that might not be an issue

**Comment by beniwohli** [comment]
@ciarancourtney oh, you're right. Weirdly enough, I can't replicate the issue in Django 1.11 with runserver

File: elasticapm/transport/base.py (hunk 1 above)
Comment (simitt) [reply elasticapm/transport/base.py]: agree, this shouldn't really matter

**Comment by beniwohli** [comment]
merged in 4a02ee52a8f8f0ae63f97c9abc00dd1f2a9b527e

**Pull request closed without merging**